hoomd/mpcd/CosineChannelGeometry.h

```diff
@@ -0,0 +1,317 @@
+                                - ((m_amplitude * fast::cos(x0 * m_wavenumber) + sign * m_H)
+                                   - vel.y / vel.x * (x0 - pos.x) - pos.y));
+
+            Scalar n, n2;
+            Scalar s, c;
+            unsigned int counter = 0;
+            while (delta > target_precision && counter < max_iteration)
+                {
+                fast::sincos(x0 * m_wavenumber, s, c);
+                n = (m_amplitude * c + sign * m_H) - vel.y / vel.x * (x0 - pos.x) - pos.y; // f
+                n2 = -m_wavenumber * m_amplitude * s - vel.y / vel.x;                      // df
+                x0 = x0 - n / n2; // x = x - f/df
+                delta = fabs(0
+                             - ((m_amplitude * fast::cos(x0 * m_wavenumber) + sign * m_H)
+                                - vel.y / vel.x * (x0 - pos.x) - pos.y));
+                ++counter;
+                }
+
+            /* The new z position is calculated from the wall equation to guarantee that the new
+             * particle positon is exactly at the wall and not accidentally slightly inside of the
+             * wall because of nummerical presicion.
+             */
+            y0 = (m_amplitude * fast::cos(x0 * m_wavenumber) + sign * m_H);
+
+            /* The new y position can be calculated from the fact that the last position outside of
+             * the wall, the current position inside of the  wall, and the new position exactly at
+             * the wall are on a straight line.
+             */
+            z0 = -(pos.x - dt * vel.x - x0) * vel.z / vel.x + (pos.z - dt * vel.z);
+
+            // Newton's method sometimes failes to converge (close to saddle points, df'==0, bad
+            // initial guess,overshoot,..) catch all of them here and do bisection if Newthon's
+            // method didn't work
+            Scalar lower_x = fmin(pos.x - dt * vel.x, pos.x);
+            Scalar upper_x = fmax(pos.x - dt * vel.x, pos.x);
+
+            // found intersection is NOT in between old and new point, ie intersection is
+            // wrong/inaccurate. do bisection to find intersection - slower but more robust than
+            // Newton's method
+            if (x0 < lower_x || x0 > upper_x)
+                {
+                counter = 0;
+                Scalar3 point1 = pos;            // final position at t+dt, outside of channel
+                Scalar3 point2 = pos - dt * vel; // initial position, inside of channel
+                Scalar3 point3 = 0.5 * (point1 + point2); // halfway point
+                Scalar fpoint3 = (m_amplitude * fast::cos(point3.x * m_wavenumber) + sign * m_H)
+                                 - point3.y; // value at halfway point, f(x)
+                // Note: technically, the presicion of Newton's method and bisection is slightly
+                // different, with bisection being less precise and slower convergence.
+                while (fabs(fpoint3) > target_precision && counter < max_iteration)
```

Comment:
Is this still how these converge checks should be handled then?

Reply:
I think so yes. This is tested on various reasonable geometries, and works in cases where Newton's method fails.